### ISSUE TITLE
Replacing PoseCNN with actual Pose

### ISSUE THREAD

**Issue by aj96** [nianticlabs/monodepth2#70]
Has anyone tried to replace the PoseCNN with the actual cam2cam pose? I know that the kitti dataset provides the pose as part of the oxts folder. They provide lat, lon, alt, roll, pitch, and yaw as described in the readme here: https://github.com/pratikac/kitti/blob/master/readme.raw.txt 

I know that to create that actual cam2cam transformation I'd need to chain together quite a few transformations cam_t0 -> velo_t0 -> imu_t0 -> imu_t1 -> velo_t1 -> cam_t1. But I haven't been able to get it quite right. Currently I am converting the lat, lon, alt to ECEF and composing the 6-dof information into a 4x4 transformation matrix which I call local vertical to imu. But I know it is wrong because it is saying the camera moves hundreds of meters between one time step. After doing a point trace, the problem is unsurprisingly occurring during the imu_t0 -> imu_t1 transformation.

Can anyone explain how to use the provided lat, lon, alt, roll, pitch, and yaw to provide the transformation between two timesteps within imu frame? Kitti directly provides velo2cam and imu2velo, so those are not an issue.

Thank you.

**Comment by mrharicot** [comment]
If I remember correctly the pose in the raw data isn't precise enough to use for reprojection.
I would try to visualize the corresponding path in 
3D with matplotlib first.
Let us know what you find out!

**Comment by aj96** [comment]
Can you direct me to a source for where you got that information?

And sure, I will plot the data and let you guys know.

**Comment by mrharicot** [comment]
My bad, I went back and checked the [original KITTI paper](http://www.cvlibs.net/publications/Geiger2012CVPR.pdf) and the ground truth odometry is taken from the GPS/IMU unit.

**Comment by yuqi1991** [comment]
@aj96 hi there,
    I saw your comments over the training with real cam to cam pose, actually I'm trying to train my custom dataset with real pose just like you, but it always result in some weird outputs like all zero disparity map.  I have disabled the augmented flipping image option and I'm sure my real pose data from imu is accurate. Could you share me what other things should I change to get the correct depth map ?

**Comment by aj96** [comment]
@yuqi1991 Hey, it's actually a super tedious and difficult task. 

You need to compose this transformation: 
(velo2cam) * (imu2velo) * (enu0_2_imu_t') * (imu_t_2_enu0) * (velo2imu) * (cam2velo)

This gives you your cam2cam transformation from one time-step t to another time-step t'

To get the translation vector for imu2enu, you can use this code:
```
import pymap3d as pm
tx, ty, tz = pm.geodetic2enu(lat_t, lon_t, alt_t, lat_ref, lon_ref, alt_ref)

```
Your reference lat, lon, alt should be from time-step 0, which obviously changes depending on which temporal sequence you are using.

To get R_imu_t_2enu0, you need to compose these three rotation matrices together: (R_ecef_2_enu0) * (R_enut_2_ecef) * (R_imut_2_enut). Then you can compose your translation vector and rotation matrix into your 4x4 (imu_t_2_enu0). Just take the inverse to get the corresponding (enu0_2_imu_t) matrix.

This repo has the code to get those rotation matrices: https://github.com/sameeptandon/sail-car-log/blob/master/process/WGS84toENU.py

One more thing, there are some noisy examples in the kitti imu data. Some of the readings are giving a change in altitude of as high as 5 meters between one time-step. I plotted the six-dof information and created a new split I called eigen_zhou_localization which took out any examples that had obvious outliers.

The model doesn't do as well as with posecnn, which is expected because you halved the number of learnable parameters, but it still learns well enough and to metric scale. There was only a difference of 0.1 abs_rel error.

I hope this was helpful!

**Comment by yuqi1991** [comment]
@aj96 Thanks for your reply, I wonder does it need to change the ‘disp_to_depth’ converter as well, since the scaling factor can be retrieve by the real pose.

**Comment by aj96** [comment]
@yuqi1991 I didn't, and it worked fine, so I don't think it's needed.

**Comment by mrharicot** [comment]
> The model doesn't do as well as with posecnn, which is expected because you halved the number of learnable parameters,

While in general halving the number of parameters hurts performance, in this case the model should actually do significantly better than with poseCNN as it is using "ground truth" poses.
I believe the loss in accuracy could come from noisy odometry.

**Comment by aj96** [comment]
@mrharicot That makes sense. @yuqi1991 If you get better performance than using posecnn, do let us know!

**Comment by yuqi1991** [comment]
@aj96 hi, yesterday I finally get the results using my real pose, the abs_rel error is about 0.7, I think it's due to lack of training samples (I got almost 200 sets of photos). I will try again with more training data  and replacing ResNet18 with bigger models

**Comment by WajahatAkhtar** [comment]
Hello @yuqi1991 ! 
Can you share the repository/script where you manage to change the posecnn with G-T pose from kitti. As, I am also working on a similar issue, but not getting good results, might be it can help me to find where i am wrong.

**Comment by aditbhrgv** [comment]
> Hello @yuqi1991 !
> Can you share the repository/script where you manage to change the posecnn with G-T pose from kitti. As, I am also working on a similar issue, but not getting good results, might be it can help me to find where i am wrong.

Hi,

Have you managed to get the script ? thanks

**Comment by dexter2406** [comment]
> @aj96 hi, yesterday I finally get the results using my real pose, the abs_rel error is about 0.7, I think it's due to lack of training samples (I got almost 200 sets of photos). I will try again with more training data and replacing ResNet18 with bigger models

Same here. Could you share the snippet?

**Comment by dexter2406** [comment]
@yuqi1991 Could you share the snippet? Thanks
